game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,125 @@
+package games.strategy.triplea.delegate.battle.steps.fire;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.delegate.Matches;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Group of units that are firing on targets
+ *
+ * <p>If the group is suicideOnHit, then all of the units should have the same unit type
+ */
+@Value
+public class FiringGroup {
+
+  private static final UnitType NON_SUICIDE_MULTIMAP_KEY = new UnitType("nonsuicide", null);
+
+  String displayName;
+  String groupName;
+  Collection<Unit> firingUnits;
+  Collection<Unit> targetUnits;
+  boolean suicideOnHit;
+
+  private FiringGroup(
+      final String displayName,
+      final String groupName,
+      final Collection<Unit> firingUnits,
+      final Collection<Unit> targetUnits) {
+    this.displayName = displayName;
+    this.groupName = groupName;
+    this.firingUnits = firingUnits;
+    this.targetUnits = targetUnits;
+    this.suicideOnHit = this.firingUnits.stream().allMatch(Matches.unitIsSuicideOnHit());
+  }
+
+  public Collection<Unit> getTargetUnits() {
+    return Collections.unmodifiableCollection(targetUnits);
+  }
+
+  public Collection<Unit> getFiringUnits() {
+    return Collections.unmodifiableCollection(firingUnits);
+  }
+
+  /**
+   * Keeps alive units around
+   *
+   * <p>Units unfortunately don't track their own status so the list of targets needs to be updated
+   * as the battle progresses.
+   *
+   * @param aliveUnits Units that are still alive
+   */
+  public void retainAliveTargets(final Collection<Unit> aliveUnits) {
+    targetUnits.retainAll(aliveUnits);
+  }
+
+  /**
+   * Splits up the firingUnits by suicideOnHit status and groups them by unit type
+   *
+   * @param name Name of the firing units
+   * @param firingUnits Collection of units that are firing
+   * @param targetUnits Collection of units that are being hit
+   * @return List of FiringGroup
```

Comment:
nit, not sure this @return javadoc adds much, restates the code. We are not publishing library code, anyone reading the javadoc will likely be reading the javadoc in the class file at the same time. I think this `@return` can perhaps be removed or otherwise should describe any invariants that would be true of the return value. For example, it is a list, what is the ordering of the list? Are there constraints on the size of the list. Also, will the firigin group be a collection of the firing units, or the target units, or both?

Reply:
I am a bit confused though, `suicideOnHit` already implies a specific unit type, right? If we are grouping by unit type, then don't we automatically also group by `suicideOnHit` too?